policies/compatibility.rego

```diff
@@ -0,0 +1,352 @@
+# is considered a backward compatibility violation.
+#
+# In other words, we do not allow the removal of an attribute once added
+# to the registry. It must exist SOMEWHERE in a group, but may be deprecated.
+deny contains back_comp_violation(description, group_id, attr.name) if {
+    # Check if an attribute from the baseline is missing in the current registry
+    some attr in baseline_attributes
+    not registry_attribute_names[attr.name]
+
+    # Generate human readable error.
+    group_id := data.semconv.baseline_group_ids_by_attribute[attr.name]
+    description := sprintf("Attribute '%s' no longer exists in the attribute registry", [attr.name])
+}
+
+
+
+# Rule: Detect Stable Attributes moving to unstable
+#
+# This rule checks for attributes that were stable in the baseline registry
+# but are no longer stable in the current registry. Once stable, attributes
+# remain forever but may be deprecated.
+deny contains back_comp_violation(description, group_id, attr.name) if {
+     # Find stable baseline attributes in latest registry.
+     some attr in baseline_attributes
+     attr.stability == "stable"
+     some nattr in registry_attributes
+     attr.name == nattr.name
+
+     # Enforce the policy
+     attr.stability != nattr.stability
+
+     # Generate human readable error.
+     group_id := data.semconv.baseline_group_ids_by_attribute[attr.name]
+     description := sprintf("Attribute '%s' was stable, but has new stability marker", [attr.name])
+}
+
+# Rule: Detect Stable Attributes changing type
+#
+# This rule checks for attributes that were stable in the baseline registry
+# but are no longer stable in the current registry. Once stable, attributes
+# remain forever but may be deprecated.
+deny contains back_comp_violation(description, group_id, attr.name) if {
+     # Find stable baseline attributes in latest registry.
+     some attr in baseline_attributes
+     attr.stability == "stable"
+     some nattr in registry_attributes
+     attr.name == nattr.name
+
+     # Enforce the policy
+     # TODO - deal with enum type changes, probably in enum sections
+     not is_enum(attr)
+     attr.type != nattr.type
+
+     # Generate human readable error.
+     group_id := data.semconv.baseline_group_ids_by_attribute[attr.name]
+     attr_type_string := type_string(attr)
+     nattr_type_string := type_string(nattr)
+     description := sprintf("Attribute '%s' was '%s', but has new type '%s'", [attr.name, attr_type_string, nattr_type_string])
+}
+
+# Rule: Detect Stable enum Attributes changing type
+#
+# This rule checks for attributes that were stable in the baseline registry
+# but are no longer stable in the current registry. Once stable, attributes
+# remain forever but may be deprecated.
+deny contains back_comp_violation(description, group_id, attr.name) if {
+     # Find stable baseline attributes in latest registry.
+     some attr in baseline_attributes
+     attr.stability == "stable"
+     some nattr in registry_attributes
```

Comment:
This is one of the instances where rewriting `registry_attributes` as shown below will accelerate the rules. Instead of having a list of attributes, we rely on a map of attribute name -> attribute.

```
registry_attributes := [name: attr |
    some g in data.semconv.registry_groups
    some attr in g.attributes
    name := attr.name
]
```